File: lib/src/services/translation/providers/wiktionary_provider.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../translation_models.dart';
import '../translation_provider.dart';

/// Wiktionary REST definition endpoint — keyless fallback for definitions.
///
/// `GET https://<lang>.wiktionary.org/api/rest_v1/page/definition/<word>`
/// Returns an object keyed by language code; definitions are HTML and must be
/// sanitized. Wikimedia etiquette requires a descriptive User-Agent.
class WiktionaryProvider implements DefinitionProvider {
  WiktionaryProvider(this._client);

  final http.Client _client;

  static const _userAgent =
      'eng-reader/1.0 (Flutter language-learning reader; contact: in-app)';

  @override
  String get id => 'wiktionary';

  @override
  String get name => 'Wiktionary';

  @override
  Future<DefinitionResult> define({
    required String word,
    required String lang,
  }) async {
    final uri = Uri.https(
      '$lang.wiktionary.org',
      '/api/rest_v1/page/definition/${Uri.encodeComponent(word)}',
    );
    final http.Response resp;
    try {
      resp = await _client.get(
        uri,
        headers: {'User-Agent': _userAgent, 'Accept': 'application/json'},
      );
    } catch (e) {
      throw ProviderException('Network error contacting Wiktionary: $e');
    }
    if (resp.statusCode == 404) {
      throw ProviderException(
        'No Wiktionary entry for "$word".',
        notFound: true,
      );
    }
    if (resp.statusCode != 200) {
      throw ProviderException('Wiktionary returned HTTP ${resp.statusCode}');
    }
    final data = jsonDecode(utf8.decode(resp.bodyBytes));
    if (data is! Map) {
      throw ProviderException(
        'No Wiktionary entry for "$word".',
        notFound: true,
      );
    }
    // Prefer definitions in the requested language; fall back to any present.
    final List<dynamic>? entries =
        (data[lang] as List?) ??
        (data.values.firstWhere((v) => v is List, orElse: () => null) as List?);
    if (entries == null || entries.isEmpty) {
      throw ProviderException(
        'No Wiktionary entry for "$word".',
        notFound: true,
      );
    }

    final senses = <DefinitionSense>[];
    for (final e in entries) {
      if (e is! Map) continue;
      final pos = (e['partOfSpeech'] as String?)?.trim() ?? '';
      final defs = e['definitions'];
      final items = <DefinitionItem>[];
      if (defs is List) {
        for (final d in defs) {
          if (d is! Map) continue;
          final def = _stripHtml((d['definition'] as String?) ?? '');
          if (def.isEmpty) continue;
          String? example;
          final examples = d['examples'];
          if (examples is List && examples.isNotEmpty) {
            example = _stripHtml(examples.first.toString());
          }
          items.add(DefinitionItem(definition: def, example: example));
        }
      }
      if (items.isNotEmpty) {
        senses.add(DefinitionSense(partOfSpeech: pos, items: items));
      }
    }
    if (senses.isEmpty) {
      throw ProviderException(
        'No Wiktionary entry for "$word".',
        notFound: true,
      );
    }
    return DefinitionResult(
      word: word,
      senses: senses,
      providerId: id,
      attribution: 'Definitions: Wiktionary (CC BY-SA)',
    );
  }

  static final _tagPattern = RegExp(r'<[^>]*>');

  /// Strip HTML tags and decode the handful of entities Wiktionary emits.
  static String _stripHtml(String html) {
    var text = html.replaceAll(_tagPattern, '');
    text = text
        .replaceAll('&amp;', '&')
        .replaceAll('&lt;', '<')
        .replaceAll('&gt;', '>')
        .replaceAll('&quot;', '"')
        .replaceAll('&#39;', "'")
        .replaceAll('&nbsp;', ' ');
    return text.trim();
  }
}

File: ios-native/Eng/Services/Providers/WiktionaryProvider.swift
import Foundation

/// Wiktionary REST definition endpoint — keyless fallback for definitions.
/// `GET https://<lang>.wiktionary.org/api/rest_v1/page/definition/<word>`
/// Returns an object keyed by language code; definitions are HTML and sanitized.
struct WiktionaryProvider: DefinitionProvider {
    var id: String { "wiktionary" }
    private static let userAgent = "eng-reader-ios/1.0 (language-learning reader; contact: in-app)"

    func define(word: String, lang: String) async throws -> DefinitionResult {
        let encoded = word.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? word
        guard let url = URL(string: "https://\(lang).wiktionary.org/api/rest_v1/page/definition/\(encoded)") else {
            throw ProviderError("Bad word", notFound: true)
        }
        let (data, resp): (Data, HTTPURLResponse)
        do { (data, resp) = try await Http.get(url, headers: ["User-Agent": Self.userAgent, "Accept": "application/json"]) }
        catch { throw ProviderError("Network error contacting Wiktionary: \(error.localizedDescription)") }
        if resp.statusCode == 404 { throw ProviderError("No Wiktionary entry for \"\(word)\".", notFound: true) }
        guard resp.statusCode == 200 else { throw ProviderError("Wiktionary returned HTTP \(resp.statusCode)") }

        guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw ProviderError("No Wiktionary entry for \"\(word)\".", notFound: true)
        }
        // Prefer definitions in the requested language; fall back to any present.
        let entries = (obj[lang] as? [[String: Any]]) ?? (obj.values.first { $0 is [[String: Any]] } as? [[String: Any]])
        guard let entries, !entries.isEmpty else {
            throw ProviderError("No Wiktionary entry for \"\(word)\".", notFound: true)
        }

        var senses: [DefinitionSense] = []
        for e in entries {
            let pos = (e["partOfSpeech"] as? String)?.trimmingCharacters(in: .whitespaces) ?? ""
            var items: [DefinitionItem] = []
            if let defs = e["definitions"] as? [[String: Any]] {
                for d in defs {
                    let def = Self.stripHtml((d["definition"] as? String) ?? "")
                    if def.isEmpty { continue }
                    var example: String?
                    if let examples = d["examples"] as? [Any], let first = examples.first {
                        example = Self.stripHtml("\(first)")
                    }
                    items.append(DefinitionItem(definition: def, example: example))
                }
            }
            if !items.isEmpty { senses.append(DefinitionSense(partOfSpeech: pos, items: items)) }
        }
        if senses.isEmpty { throw ProviderError("No Wiktionary entry for \"\(word)\".", notFound: true) }
        return DefinitionResult(word: word, senses: senses, providerId: id,
                                attribution: "Definitions: Wiktionary (CC BY-SA)")
    }

    private static let tagPattern = try! NSRegularExpression(pattern: "<[^>]*>")

    /// Strip HTML tags and decode the handful of entities Wiktionary emits.
    static func stripHtml(_ html: String) -> String {
        let ns = html as NSString
        var text = tagPattern.stringByReplacingMatches(
            in: html, range: NSRange(location: 0, length: ns.length), withTemplate: "")
        let entities = ["&amp;": "&", "&lt;": "<", "&gt;": ">", "&quot;": "\"", "&#39;": "'", "&nbsp;": " "]
        for (k, v) in entities { text = text.replacingOccurrences(of: k, with: v) }
        return text.trimmingCharacters(in: .whitespacesAndNewlines)
    }
}
